[BDCE] Handle multi-use `and`/`or`/`xor` on demanded bits

antoniofrighetto opened the pull request:
Simplify multi-use `and` or `or`s when these last do not affect the demanded bits being considered.

Fixes: https://github.com/llvm/llvm-project/issues/78596.

Proofs: https://alive2.llvm.org/ce/z/kmVqE-.

dtcxzyw commented on llvm/lib/Transforms/Scalar/BDCE.cpp:
```diff
@@ -125,6 +125,33 @@ static bool bitTrackingDCE(Function &F, DemandedBits &DB) {
+      APInt Demanded = DB.getDemandedBits(SI);
+      if (!Demanded.isAllOnes()) {
+        for (Use &U : I.operands()) {
+          auto *And = dyn_cast<BinaryOperator>(&U);
+          if (!And || !And->hasOneUse() || And->getOpcode() != Instruction::And)
+            continue;
+
+          auto *CV = dyn_cast<ConstantInt>(And->getOperand(1));
+          if (!CV)
+            continue;
```
```suggestion
          const APInt* Mask;
          if (!match(U, m_OneUse(m_And(m_Specific(&I), m_APInt(Mask)))))
            continue;
```

dtcxzyw commented on llvm/lib/Transforms/Scalar/BDCE.cpp:
```diff
@@ -125,6 +125,33 @@ static bool bitTrackingDCE(Function &F, DemandedBits &DB) {
+    if (SelectInst *SI = dyn_cast<SelectInst>(&I)) {
+      APInt Demanded = DB.getDemandedBits(SI);
+      if (!Demanded.isAllOnes()) {
+        for (Use &U : I.operands()) {
```
```suggestion
        for (Use &U : I.operands().drop_front()) {
```

dtcxzyw commented on llvm/lib/Transforms/Scalar/BDCE.cpp:
```diff
@@ -125,6 +125,33 @@ static bool bitTrackingDCE(Function &F, DemandedBits &DB) {
+            continue;
+
+          APInt Mask = CV->getValue();
+          if ((Mask | ~Demanded).isAllOnes()) {
```
```suggestion
          if (Demanded.isSubsetOf(Mask)) {
```
It would be clearer than the original.

dtcxzyw commented on llvm/lib/Transforms/Scalar/BDCE.cpp:
```diff
@@ -125,6 +125,33 @@ static bool bitTrackingDCE(Function &F, DemandedBits &DB) {
+            clearAssumptionsOfUsers(And, DB);
+            U.set(And->getOperand(0));
+            Worklist.push_back(And);
+            Changed = true;
```
```suggestion
            ++NumSimplified;
            Changed = true;
```

antoniofrighetto replied:
Thanks, must have gone away accidentally while rebasing.

nikic reviewed (changes requested):
...wat? This should not be looking at selects. You should only be looking at the demanded bits of the and.

antoniofrighetto commented:
Could you kindly provide an example showing how not restricting this to the `select` may be beneficial? We don't need to specialize it for the `select`, but restricting the input space looked reasonable to me.

> You should only be looking at the demanded bits of the and.

You mean we should only be looking at the demanded bits of the instruction itself when the operand analyzed is the `and`? I don't see how the optimization can take place if the instruction examined is not the multi-use instruction.

nikic commented:
> Could you kindly provide an example showing how not restricting this to the `select` may be beneficial? We don't need to specialize it for the `select`, but restricting the input space looked reasonable to me.

You can replace the select with pretty much any instruction that transfers demanded bits, e.g. https://alive2.llvm.org/ce/z/om-5yP.

> 
> > You should only be looking at the demanded bits of the and.
> 
> You mean we should only be looking at the demanded bits of the instruction itself when the operand analyzed is the `and`? I don't see how the optimization can take place if the instruction examined is not the multi-use instruction.

I would expect you to query the demanded bits of the `and` instruction, similar to the code directly above that handles `sext`. Unless I'm missing something, that will cover the motivating pattern.

XChy commented:
For a single BDCE pass, I think a minimal testcase can be like:
```llvm
define void @src(i64 %a) {
entry:
  %and = and i64 %a, 24
  %ret1 = and i64 %and, 8
  %ret2 = and i64 %and, 16
  call void @use(i64 %ret1)
  call void @use(i64 %ret2)
  ret void
}

define void @tgt(i64 %a) {
entry:
  %ret1 = and i64 %a, 8
  %ret2 = and i64 %a, 16
  call void @use(i64 %ret1)
  call void @use(i64 %ret2)
  ret void
}
```

antoniofrighetto committed:
[BDCE] Introduce test for PR79688 (NFC)

antoniofrighetto committed:
[BDCE] Handle multi-use binary ops upon demanded bits

Simplify multi-use `and`/`or`/`xor` when these last
do not affect the demanded bits being considered.

Fixes: https://github.com/llvm/llvm-project/issues/78596.

Proofs: https://alive2.llvm.org/ce/z/EjuWHa.

antoniofrighetto commented on llvm/lib/Transforms/Scalar/BDCE.cpp:
```diff
@@ -125,6 +125,33 @@ static bool bitTrackingDCE(Function &F, DemandedBits &DB) {
+      if (!Demanded.isAllOnes()) {
+        for (Use &U : I.operands()) {
+          auto *And = dyn_cast<BinaryOperator>(&U);
+          if (!And || !And->hasOneUse() || And->getOpcode() != Instruction::And)
```
Sure, thanks, adding it too.

antoniofrighetto commented:
> I would expect you to query the demanded bits of the and instruction, similar to the code directly above that handles sext. Unless I'm missing something, that will cover the motivating pattern.

Reviewed, thanks, somehow I thought it would be easier to handle the `and`/`or` as operands of the instruction being analyzed (and get its current demanded bits).

XChy commented on llvm/lib/Transforms/Scalar/BDCE.cpp:
```diff
@@ -125,6 +125,28 @@ static bool bitTrackingDCE(Function &F, DemandedBits &DB) {
+      if (Opc == Instruction::And || Opc == Instruction::Or) {
+        APInt Demanded = DB.getDemandedBits(BO);
+        if (!Demanded.isAllOnes()) {
+          if (auto *CV = dyn_cast<ConstantInt>(BO->getOperand(1))) {
+            APInt Mask = CV->getValue();
+            if ((Opc == Instruction::And && Demanded.isSubsetOf(Mask)) ||
+                (Opc == Instruction::Or && !Demanded.intersects(Mask))) {
```
```suggestion
          const APInt *Mask;
          if (match(BO->getOperand(1), m_APInt(Mask))) {
            if ((Opc == Instruction::And && Demanded.isSubsetOf(*Mask)) ||
                (Opc == Instruction::Or && !Demanded.intersects(*Mask))) {
```
Match m_APInt to handle the splat.

XChy commented on llvm/lib/Transforms/Scalar/BDCE.cpp:
```diff
@@ -125,6 +125,28 @@ static bool bitTrackingDCE(Function &F, DemandedBits &DB) {
+    // Simplify `and` or `or` when their mask does not affect the demanded bits.
+    if (auto *BO = dyn_cast<BinaryOperator>(&I)) {
+      unsigned Opc = BO->getOpcode();
+      if (Opc == Instruction::And || Opc == Instruction::Or) {
```
Could you turn it into a switch, please? I think there is room for other operators like `xor`, `add` to be here in the future.

nikic commented on llvm/test/Transforms/BDCE/and-or-multiuse.ll:
```diff
@@ -0,0 +1,204 @@
+entry:
+  %and1 = and i64 %a, 24                          ; Mask:          0001 1000
+  %and2 = and i64 %b, 25                          ; Mask:          0001 1001
+  %or = or i64 %and1, %and2
```
Can you please add a test variant with `or disjoint`, to show that the flag gets dropped correctly?

nikic commented on llvm/test/Transforms/BDCE/dead-uses.ll:
```diff
@@ -9,8 +9,7 @@ declare <2 x i32> @llvm.fshr.v2i32(<2 x i32>, <2 x i32>, <2 x i32>)
 ; First fshr operand is dead.
 define i32 @pr39771_fshr_multi_use_instr(i32 %a) {
 ; CHECK-LABEL: @pr39771_fshr_multi_use_instr(
-; CHECK-NEXT:    [[X:%.*]] = or i32 [[A:%.*]], 0
```
Can you change the `or` operand to  `1` (or something else that doesn't optimize away) instead? Dropping it is not really the point of this tests.

nikic reviewed (commented):
Thanks, this looks a lot more like what I had in mind.

antoniofrighetto commented:
Reviewed, thanks.

nikic reviewed (approved):
LGTM

XChy reviewed (approved):
LGTM.
